File: 백준/Gold/6443. 애너그램/애너그램.java
import java.io.IOException;
import java.util.*;

public class Main {
    static char[] words;
    static List<String> result;
    static int[] alpha = new int[26];
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.nextLine();
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < n; i++) {
            result = new ArrayList<>();
            alpha = new int[26];
            words = sc.nextLine().toCharArray();
            for(char c : words){
                int index = c - 'a';
                alpha[index]++;
            }
            dfs(0, new StringBuilder());
            for (String s : result) {
                sb.append(s).append("\n");
            }
        }
        System.out.println(sb);
    }
    static void dfs(int depth, StringBuilder sb){
        if(depth== words.length){
            result.add(sb.toString());
            return;
        }
        for(int i = 0; i < 26; i++){
            if(alpha[i] != 0){
                char index = (char) (i+ 'a');
                alpha[i]--;
                sb.append(index);
                dfs(depth+1, sb);
                sb.deleteCharAt(sb.length()-1);
                alpha[i]++;
            }
        }
    }
}

File: 백준/Gold/6443. 애너그램/애너그램.kt
import java.util.*


fun main(){
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    sc.nextLine()
    val sb = StringBuilder()
    for(i in 0 until n){
        result = ArrayList()
        alpha = IntArray(26)
        words = sc.nextLine().toCharArray()
        for(c in words){
            var index = c - 'a'
            alpha[index]++
        }
        dfs(0, StringBuilder())
        for(s in result){
            sb.append(s).append("\n")
        }
    }
    println(sb)
}

lateinit var words: CharArray
lateinit var result: MutableList<String>
var alpha = IntArray(26)

fun dfs(depth: Int, sb:StringBuilder){
    if(depth == words.size){
        result.add(sb.toString())
        return
    }
    for(i in 0..25){
        if(alpha[i] != 0){
            val index = (i + 'a'.code).toChar()
            alpha[i]--
            sb.append(index)
            dfs(depth+1, sb)
            sb.deleteCharAt(sb.length-1)
            alpha[i]++
        }
    }
}